[clang-tidy] Update TODO comment check

AlanRosenthal opened the pull request:
The doc for google-readability-todo reference lists two styles of TODO comments:
https://google.github.io/styleguide/cppguide.html#TODO_Comments

Previously, only `TODO(info): comment` were supported.
After this PR, `TODO: info - comment` are also supported.

njames93 commented on clang-tools-extra/clang-tidy/google/TodoCommentCheck.cpp:
```diff
@@ -45,7 +54,7 @@ class TodoCommentCheck::TodoCommentHandler : public CommentHandler {
   TodoCommentCheck &Check;
   std::string User;
-  llvm::Regex TodoMatch;
+  std::list<std::string> TodoMatches;
```
This is horribly inefficient, if you need multiple regexes use a vector of `llvm::regex`.
However for this patch there doesn't need to be 2 regexes as you can match both cases with just the one

AlanRosenthal replied:
Hi @njames93, thanks for the review. I've updated the PR to initialize the regexes in the constructor.

I had thought about combining the regex, but I had concluded that it would make it very difficult to read / update / maintain.

Please take a look at the current version, thanks!

AlanRosenthal committed:
[clang-tidy] Update TODO comment check

The doc for google-readability-todo reference:
https://google.github.io/styleguide/cppguide.html#TODO_Comments
Which lists two styles of TODO comments.

Previously, only `TODO(John): comment` were supported.
After this PR, `TODO: bug 123 - comment` are also supported.

5chmidti commented on clang-tools-extra/clang-tidy/google/TodoCommentCheck.cpp:
```diff
@@ -9,6 +9,7 @@
 #include "TodoCommentCheck.h"
 #include "clang/Frontend/CompilerInstance.h"
 #include "clang/Lex/Preprocessor.h"
+#include <list>
```
Please remove this unused include

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/google/readability-todo.cpp:
```diff
@@ -24,3 +24,7 @@
 // TODO(b/12345): find the holy grail
 // TODO (b/12345): allow spaces before parentheses
 // TODO(asdf) allow missing semicolon
+// TODO: bug 12345678 - Remove this after the 2047q4 compatibility window expires.
+// TODO: example.com/my-design-doc - Manually fix up this code the next time it's touched.
+// TODO(bug 12345678): Update this list after the Foo service is turned down.
+// TODO(John): Use a "\*" here for concatenation operator.
```
Given that this is the text from the style-guide, I'd be concerned regarding licensing? It's under CC 3.0
You could just rewrite the lines that were already here to fit the pattern

AlanRosenthal replied:
Great point, done.

AlanRosenthal committed:
PR feedback

AlanRosenthal committed:
fix typo

5chmidti commented on clang-tools-extra/test/clang-tidy/checkers/google/readability-todo.cpp:
```diff
@@ -24,3 +24,5 @@
 // TODO (b/12345): allow spaces before parentheses
 // TODO(asdf) allow missing semicolon
+// TODO: b/12345 - solve the collatz conjecture
+// TODO: foo - bar
```
Please add a case for both forms: `TODO(foo): bar - baz`

5chmidti reviewed (commented):
(noticed I had one pending review comment still open)

Please add a release note (see https://github.com/llvm/llvm-project/blob/412e3e394dbd1b7d8655639e161ed4dbd5505c96/clang-tools-extra/docs/ReleaseNotes.rst?plain=1#L104-L105)
